docs/references/architecture/adr-115-predictable-block-times.md

```diff
@@ -0,0 +1,133 @@
+network. You instruct validators in your network to change `timeout_commit`
+(_how long a node waits for additional precommits after committing a block,
+before starting on the next height_) from 1s to 4s in the node's config file.
+
+Do you have predictable block times now?
+
+No - you don't. The expected block time will be around 6s, but even in
+favorable runs it will drift far apart from 6s due to:
+
+1. CometBFT going into multiple rounds of consensus (it happens rarely but
+   still).
+2. Network latency.
+3. Clock drifts.
+4. The delay for processing a block.
+
+A validator could also change `timeout_commit` to 0s in its node's config file.
+That means whenever THIS validator proposes a block, the block time will be 2s
+(not 6s!).
+
+Note that the value of `timeout_commit` is static and can't be changed
+without restarting the node.
+
+Because 1-3 are out of your (and our) control, **we can't have constant block
+times**. But we can design a mechanism so that the medium-to-long term average
+block time converges to a desired value.
+
+To achieve that, we need to define a form of variable block time. Namely, if a
+block takes longer than expected, we should be able to render the next block(s)
+faster in order to converge to the desired (average) block time.
+
+### Use Cases
+
+In case of Osmosis, committing a big block is expected to take longer than
+usual. If we know the approximate size of the next block, we could increase
+`timeout_commit` dynamically (if such feature existed) to give the state
+machine some extra time to finish execution.
+
+In case of Berachain, they want to have a constant block time to match
+Ethereum's slots, which are equal to 12s.
+
+Celestia's usecase here for the medium term is to be able to have a longer
+proposal timeout so that they can spend a larger percentage of the block time
+gossiping while having consistent block times.
+
+## Proposal
+
+Move `timeout_commit` into `FinalizeBlockResponse` as `next_block_delay`. This
+field's semantics stays essentially the same: delay between the time when the
+current block is committed and the next height is started. The idea is
+literally to have the same behavior as `timeout_commit`, while allowing the
+application to pick a different value for each height.
+
+If Comet goes into multiple rounds of consensus, the ABCI application can react
+by lowering `next_block_delay`. Of course, nobody could guarantee that there
+won't be 100000 rounds of consensus, so **it's still best effort** when it
+comes to individual block times.
+
+## Alternative Approaches
+
+1. Do nothing. The block time will stay largely unpredictable.
+2. Make `timeout_commit` global (= consensus parameter). It doesn't solve the
+   problem of multiple rounds of consensus + the execution is delayed by one
+   block (as with other consensus parameters).
+3. Add `next_block_delay`, but keep `timeout_commit`. Individual validators can
```

Comment:
I think we'd need to keep `timeout_commit` around for at least one more release (but we deprecate it now). The reason is UX of upgrade. If an appchain doesn't care about this at the time of upgrading, they shouldn't _feel_ any pain.
A possible way of achieving this would be:
* keep `timeout_commit` deprecated
* if a value for `timeout_commit` is detected at process start-up, warn the user (WARNING log?) that they are using a deprecated field
* if the app provides a value for `next_block_delay`, then `timeout_commit` is ignored
* if the app does not provide a value for `next_block_delay`, then CometBFT falls back to `timeout_commit`

Reply:
Good proposal 👍